streaming/base/storage/upload.py

```diff
@@ -0,0 +1,330 @@
+            progress_bar (bool): Whether to display a progress bar while uploading to a cloud
+                provider. Defaults to ``False``.
+
+        Raises:
+            ValueError: Either local and/or remote path(s) must be provided
+            KeyError: Invalid Cloud provider prefix
+
+        Returns:
+            _type_: _description_
+        """
+        if not local and not remote:
+            raise ValueError('You must provide local and/or remote path(s).')
+
+        mapping = {
+            's3': S3Writer,
+            'gs': GCSWriter,
+            'oci': OCIWriter,
+            '': LocalWriter,  # For local copy when remote is a local directory
+            b'': LocalWriter  # No file copy to remote since remote is None
+        }
+        obj = urllib.parse.urlparse(remote)
+        if obj.scheme in mapping:
+            instance = super().__new__(mapping[obj.scheme])
+            return instance
+        else:
+            raise KeyError('Invalid Cloud provider prefix in `remote` argument.')
+
+    def __init__(self,
+                 local: Optional[str] = None,
+                 remote: Optional[str] = None,
+                 keep_local: bool = False,
+                 progress_bar: bool = False) -> None:
+        """Initialize and validate local and remote path.
+
+        Args:
+            local (str, optional): Optional local output dataset directory. If not
+                provided, a random temp directory will be used. If ``remote`` is provided,
+                this is where shards are cached before uploading. One or both of ``local``
+                and ``remote`` must be provided. Defaults to ``None``.
+            remote (str, optional): Optional remote output dataset directory. If not
+                provided, no uploading will be done. Defaults to ``None``.
+            keep_local (bool): If the dataset is uploaded, whether to keep the local dataset
+                shard file or remove it after uploading. Defaults to ``False``.
+            progress_bar (bool): Whether to display a progress bar while uploading to a cloud
+                provider. Defaults to ``False``.
+
+        Raises:
+            ValueError: Either local and/or remote path(s) must be provided
+            FileExistsError: Local directory must be empty
+        """
+        self.keep_local = keep_local
+        self.progress_bar = progress_bar
+        if local is not None:
+            self.local = local
+            self.remote = remote
+        elif remote is not None:
+            self.local = mkdtemp()
+            self.remote = remote
+        else:
+            raise ValueError('You must provide local and/or remote path(s).')
+
+        if os.path.exists(self.local) and len(os.listdir(self.local)) != 0:
+            raise FileExistsError(f'Directory is not empty: {self.local}')
+        os.makedirs(self.local, exist_ok=True)
+
+    def upload_file(self, filename: str):
```

Comment:
we should specify that this is a relative path?

Reply:
Thanks! Addressed as part of https://github.com/mosaicml/streaming/pull/171/commits/5b5c10450637d4ccdbc8b5c2403a7d18f89b169b.